### PR TITLE
feat: ✨ Advanced Mod Configs

### LINKED ISSUE #178: Expand configs: Add type, min/max
As discussed on the Brotato Discord [here](https://discord.com/channels/630060181086142487/1050470169098588232/1084090938407788544), it would be good to expand the configs to add support for setting types, and perhaps min/max values too.

It adds a bit of extra complexity when parsing the default configs, but I think it's worth it, as it would make implementing a GUI much easier for modders/game developers.

**Example:**

image

**Ideas for options:**

- All: `type`, `default`
- Text: `min_length`, `max_length`, `regex`
- Float/Int: `min_value`, `max_value`, `step` (ie. the amount the value should be incremented by in a mod's GUI)

### PR DESCRIPTION
**Pull request by KANAjetzt**
# Advanced Mod Configs - `ModLoaderConfig` and `ModConfig`

## TLDR
This PR introduces a comprehensive rework of the Mod Config setup, addressing the requirements outlined in issue #178. Key changes include the addition of `JSONSchema` Validation and the implementation of a new `Resource` Class called `ModConfig`, which defines the structure of the Config Data. The existing `ModLoaderConfig` has been completely rewritten, offering new API methods that cater to the various needs of creating a UI for editing Mod Configs and integrating them within a Mod Project.

## New API Methods `ModLoaderConfig`

- `create_config(mod_id: String, config_name: String, config_data: Dictionary) -> ModConfig`
Creates a new configuration for a mod, stores it in the mod's `ModData` and saves it to a new config JSON file in the mod's config directory.

- `update_config(config: ModConfig) -> ModConfig`
Updates an existing `ModConfig` and saves it to disk.

- `delete_config(config: ModConfig) -> bool`
 Deletes a `ModConfig`, removes the Object from the mod's `ModData` and deletes the corresponding JSON file.

- `set_current_config(config: ModConfig) -> void`
Sets the current configuration of a mod to the specified configuration.

- `get_config_schema(mod_id: String) -> Dictionary`
Returns the schema for the specified mod id.
If no configuration file exists for the mod, an empty dictionary is returned.

- `get_schema_for_prop(config: ModConfig, prop: String) -> Dictionary`
Retrieves the schema for a specific property key.
The property key is specified as a string using the `.` notation:
`"parentProp.childProp.nthChildProp"` or `"propKey"`

- `_traverse_schema(schema_prop: Dictionary, prop_key_array: Array) -> Dictionary`
Recursively traverses the schema dictionary based on the provided `prop_key_array` and returns the corresponding schema for the target property.
Used internally by `get_schema_for_prop()`

- `get_mods_with_config() -> Array`
Retrieves an Array of mods that have configuration files.

- `get_configs(mod_id: String) -> Dictionary`
Retrieves the configurations dictionary for a given mod ID.

- `get_config(mod_id: String, config_name: String) -> ModConfig`
Retrieves the configuration for a specific mod and configuration name.

- `get_default_config(mod_id: String) -> ModConfig`
Retrieves the default configuration for a specified mod ID.

- `get_current_config(mod_id: String) -> ModConfig`
Retrieves the currently active configuration for a specific mod.

- `get_current_config_name(mod_id: String) -> String`
Retrieves the name of the current configuration for a specific mod.

### Properties

- `name: String`
Name of the config - must be unique

- `mod_id: String`
The mod_id this config belongs to

- `schema: Dictionary`
The JSON-Schema this config uses for validation

- `data: Dictionary`
The data this config holds

- `save_path: String`
The path where the JSON file for this config is stored

- `is_valid := false`
False if any data is invalid

### Methods

- `get_data_as_string() -> String`
Returns the config data as a JSON String

- `get_schema_as_string() -> String`
Returns the schema as a JSON String

- `validate() -> String`
Performs the JSON-Schema validation and returns an empty string if the data is valid, or an error message if there are validation errors.

- `is_valid() -> bool`
Runs the JSON-Schema validation and returns true if valid.

- `save_to_disc() -> bool`
Saves the config data to the `save_path`.

- `remove_from_disc() -> bool`
Removes the config data from the `save_path`.

### Consts

- `CACHE_FILE_PATH`
Path to the cache file - by the time of writing this is `"user://ModLoaderCache.json"`

### Methods

- `init_cache(_ModLoaderStore) -> void`
Creates a new cache file or loads the existing one.
*ModLoaderStore is passed as parameter so the cache data can be loaded on ModLoaderStore._init()*

- `add_data(key: String, data: Dictionary) -> void`
Adds data to the cache

- `get_data(key: String) -> Dictionary`
Get data from a specific key

- `get_cache() -> Dictionary`
Get the entire cache dictionary

- `has_key(key: String) -> bool`
Check if the key exists in the cache data

- `update_data(key: String, data: Dictionary) -> Dictionary`
Updates or adds data to the cache

- `remove_data(key: String) -> void`
Remove data from the cache

- `save_to_file() -> void`
Save the cache to the cache file

- `_load_file(_ModLoaderStore = ModLoaderStore) -> void`
Load the cache file data and store it in ModLoaderStore

- `_init_cache_file() -> void`
Create an empty cache file

### New Signal 

- `current_config_changed(config)`
Emitted when the current config of a mod has changed. 
This signal can be used by mod authors to apply config changes to their mod. 
*Maybe there is a way for a mod to receive a signal specific to its config? 
This way, a mod would only be notified if its own `current_config` has changed.*

## Changes and Additions to `ModLoaderUserProfile`

The mod_list is now a Dictionary of Dictionaries.
The `current_config` for each mod is stored in the User Profile `mod_list`.

### New Methods

- `set_mod_current_config(mod_id: String, config_name: String, profile_name := ModLoaderStore.current_user_profile) -> bool`
Sets the current config for a mod in a user profiles mod_list.

- `_update_mod_list(mod_list: Dictionary, mod_data := ModLoaderStore.mod_data) -> Dictionary`
Updates the mod list by checking the validity of each mod entry and making necessary modifications.

- `_generate_mod_list() -> Dictionary`
Generates a dictionary with data to be stored for each mod.

- `_generate_mod_list_entry(mod_id: String, is_active: bool) -> Dictionary`
Generates a mod list entry dictionary with the given mod ID and active status.
If the mod has a config schema, sets the 'current_config' key to the current_config stored in the Mods ModData.

### New Properties

- `configs := {}`
Stores all `ModConfig` Objects

- `current_config: ModConfig setget _set_current_config`
Stores the currently active `ModConfig`

### New Methods

- `load_configs() -> void`
Load each mod config json from the mods config directory.

- `_load_config(config_file_path: String) -> void`
Creates a new `ModConfig` instance for each Config JSON and add it to the configs dictionary.

- `_set_current_config(new_current_config: ModConfig) -> void`
Seter function for `current_config` to update the User Profile

### Properties
- The manifest stores now the `config_schema` instead of `config_defaults`

### Methods

- `load_mod_config_defaults() -> void`
Loads the default configuration for a mod.
By loading the `default.json` config file or using `_generate_default_config_from_schema()`

- `_generate_default_config_from_schema(property: Dictionary, current_prop := {}) -> Dictionary`
Generates the default config from the `default` keys inside the Config JSON-Schema.

### _ModLoaderFile 

- `remove_file(file_path: String) -> bool`
Removes a file from the given path.

### ModLoaderUtils

 - `get_dict_from_dict(dict: Dictionary, key: String) -> Dictionary`
Returns an empty Dictionary if the key does not exist or is not type of Dictionary.

### _ModLoaderPath

 - `get_dir_paths_in_dir(src_dir_path: String) -> Array`
Returns an array of directory paths inside the src dir.

 - `get_file_paths_in_dir(src_dir_path: String) -> Array`
Returns an array of file paths inside the src dir.

 - `get_path_to_mod_configs_dir(mod_id: String) -> String`
Get the path to a mods config folder.
Returns an empty string if there is no config dir for this `mod_id`.

 - `get_path_to_mod_config_file(mod_id: String, config_name: String) -> String`
Get the path to a mods config file.
Returns an empty string if the config file does not exist.

## Known issues
- Currently, the existing Config Implementation does not have a proper deprecation process in place.
- If the user switches to a different profile and there are different current_configs for the mods saved, no current_config_changed(config) signal is fired. This results in the mods not updating their config after the profile is switched. 
*However, this is not a major issue right now because a game restart is required anyway to apply the active state of a mod.*

<br />
<br />

---

*closes #178*

### CODE REVIEW

File: addons/mod_loader/internal/path.gd [hunk 1]
```diff
@@ -105,6 +105,51 @@ static func get_flat_view_dict(p_dir := "res://", p_match := "", p_match_is_rege
 	return data
 
 
+# Returns an array of directory paths inside the src dir
+static func get_dir_paths_in_dir(src_dir_path: String) -> Array:
```
Comment (Qubus0) [inline addons/mod_loader/internal/path.gd:109]: is this used or preparation?

**Review by Qubus0 — COMMENTED** [review]
Exquisite job on this one. Thank you!
Only a few smaller things to remark, overall approved. I'll let the others have a look at it as well though

File: addons/mod_loader/internal/path.gd (hunk 1 above)
Comment (KANAjetzt) [reply addons/mod_loader/internal/path.gd:109]: That can actually be removed. Good catch👍

Comment (KANAjetzt) [reply addons/mod_loader/internal/path.gd]: fixed in 4415b363a39689a3802bb26e4ad36490f29c729e

File: addons/mod_loader/classes/mod_data.gd [hunk 2]
```diff
@@ -37,8 +37,9 @@ var is_locked := false
 var importance := 0
 # Contents of the manifest
 var manifest: ModManifest
-# Updated in _load_mod_configs
-var config := {}
+# Updated in load_mod_configs
```
[the remark this answers is not in the document]
Comment (KANAjetzt) [reply addons/mod_loader/classes/mod_data.gd]: fixed 82840f8f1a8e5c36b81655481841312a554a5581

File: addons/mod_loader/classes/mod_config.gd [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+	if not validate() == "":
+		is_valid = false
+		return false
+
+	is_valid = true
+	return true
+
+
+# Saves the config data to disc
+func save_to_disc() -> bool:
```
[the remark this answers is not in the document]
Comment (KANAjetzt) [reply addons/mod_loader/classes/mod_config.gd]: fixed 82840f8f1a8e5c36b81655481841312a554a5581

File: addons/mod_loader/classes/mod_config.gd [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+
+# Runs the JSON-Schema validation and returns true if valid
+func is_valid() -> bool:
+	if not validate() == "":
+		is_valid = false
+		return false
+
+	is_valid = true
+	return true
```
[the remark this answers is not in the document]
Comment (KANAjetzt) [reply addons/mod_loader/classes/mod_config.gd]: fixed c707437b0995ad7fbfff6e6a9f1eb826aa2faf52

File: addons/mod_loader/classes/mod_manifest.gd [hunk 5]
```diff
@@ -187,14 +189,65 @@ func to_json() -> String:
...
+		# Generate config_default based on the default values in config_schema
+		config.data = _generate_default_config_from_schema(config.schema.properties)
+		# Save the default config to disk
+		_ModLoaderFile.save_dictionary_to_json_file(config.data, config.save_path)
+	else:
+		# If there is a default.json just load that
+		config.data = _ModLoaderFile.get_json_as_dict(config.save_path)
+
+	# Validate the config defaults
+	config.is_valid()
```
[the remark this answers is not in the document]
Comment (KANAjetzt) [reply addons/mod_loader/classes/mod_manifest.gd]: We also need to update the default config if the default key in the manifest has changed. The easiest fix would be to generate the default on every start. I'm considering generating an MD5 hash of the schema and storing it somewhere. Then we can compare it and regenerate if a change is detected.

[the remark this answers is not in the document]
Comment (KANAjetzt) [reply addons/mod_loader/classes/mod_manifest.gd]: fix1 f9cd63995bc0c4c8d8c2d97518f7a40773d63ae3

File: addons/mod_loader/classes/mod_manifest.gd [hunk 6]
```diff
@@ -187,14 +189,65 @@ func to_json() -> String:
...
+
+	for property_key in property.keys():
+		var prop = property[property_key]
+
+		# If this property contains nested properties, we recursively call this function
+		if "properties" in prop:
+			current_prop[property_key] = {}
+			_generate_default_config_from_schema(prop.properties, current_prop[property_key])
+			# Return early here because a object will not have a "default" key
+			return current_prop
```
[the remark this answers is not in the document]
Comment (KANAjetzt) [reply addons/mod_loader/classes/mod_manifest.gd:243]: ty 👍 Can you open an issue for this so we can track that for 6.1?

File: addons/mod_loader/classes/mod_manifest.gd (hunk 5 above)
[the remark this answers is not in the document]
Comment (KANAjetzt) [reply addons/mod_loader/classes/mod_manifest.gd]: fix2 (`_ModLoaderCache`) f97d89d85c3712120b4f92b9ffe145eeaa304fb6

File: addons/mod_loader/internal/cache.gd [hunk 7]
```diff
@@ -0,0 +1,85 @@
+class_name _ModLoaderCache
+extends Reference
+
+
+# This Class provides methods for caching data.
+
+const CACHE_FILE_PATH = "user://ModLoaderCache.json"
```
Comment (Qubus0) [inline addons/mod_loader/internal/cache.gd:7]: probably best to still follow the snake case convention

**Review by Qubus0 — APPROVED** [review]
Very nice.
Looks like the others won't really get around to add their review, so take this approval!

File: addons/mod_loader/internal/cache.gd (hunk 7 above)
Comment (KANAjetzt) [reply addons/mod_loader/internal/cache.gd]: fixed 60c230d177f4338a48f525b40f94c08ebfc115c8
